tools/autograd/templates/Functions.cpp

```diff
@@ -717,6 +717,40 @@ Tensor diag_backward(const Tensor & grad, IntList input_sizes, int64_t diagonal)
   return grad_input;
 }
 
+Tensor diagonal_backward(const Tensor & grad, IntList input_sizes, int64_t offset, int64_t dim1_, int64_t dim2_) {
+  auto grad_input = at::zeros(grad.type(), input_sizes);
+  // the following until the assignment of auto diag
+  // copies the diagonal code in aten/src/ATen/native/TensorShape.cpp
+  // that would be equivalent to
+  //        auto diag = grad_input.diagonal(offset, dim1, dim2);
+  // when using diagonal, the output is not differentiable twice
```

Comment:
Why wouldn't it be differentiable twice if you did that? It's generally ok to use an op to implement its own backward, and there are many cases where this happens: `transpose`, `permute`, `view`, ...

Reply:
The comment documents an empirical finding - I saw the implementation of diag_backward, thought "well, I can be smart about it and use diagonal rather than manually redoing the same thing" and it failed. I changed it to do the calculations here and it worked. I do not have a theory about why it happens. :neutral_face: I don't think it's using the op that is differentiated, but it is something subtle with the inplace operation.
This is also the reason why I think it is particularly important to check that it works in the tests.

Reply:
So here we have: https://github.com/pytorch/pytorch/pull/6718/files#diff-888d69f545029e79ccc606683a7ac08cR748
Working code:
```
  auto diag = grad_input.as_strided(sizes, strides, storage_offset);
  diag.copy_(grad);
  return grad_input;
```
Non-working code:
```
  auto diag = grad_input.diagonal(offset, dim1, dim2);
  diag.copy_(grad);
  return grad_input;
```
when `aten/src/ATen/native/TensorShape.cpp`'s `diagonal` returns a `self.as_strided(...)`.
So my best guess is that the first snipped has a shortcut in the automatically generated wrappers where the second id does not. I have never looked into how ATen tracks differentiability, so I would not know.

Mind you, I think I have reproduced this several times, but I might still have done something stupid, so if someone wants to check...

Reply:
I found it! I need to register diagonal as a view function in gen_autograd.py at the top.
Thank you for insisting that I look into it - I have so much to learn.